language/vm/src/deserializer.rs

```diff
@@ -661,44 +661,137 @@ fn load_signature_tokens(cursor: &mut Cursor<&[u8]>) -> BinaryLoaderResult<Vec<S
+    // The following algorithm works by storing partially constructed types on a stack.
+    //
+    // Example:
+    //
+    //     SignatureToken: `Foo<u8, Foo<u64, bool, Bar>, address>`
+    //     Byte Stream:    Foo u8 Foo u64 bool Bar address
+    //
+    // Stack Transitions:
+    //     []
+    //     [Foo<?, ?, ?>]
+    //     [Foo<?, ?, ?>, u8]
+    //     [Foo<u8, ?, ?>]
+    //     [Foo<u8, ?, ?>, Foo<?, ?, ?>]
+    //     [Foo<u8, ?, ?>, Foo<?, ?, ?>, u64]
+    //     [Foo<u8, ?, ?>, Foo<u64, ?, ?>]
+    //     [Foo<u8, ?, ?>, Foo<u64, ?, ?>, bool]
+    //     [Foo<u8, ?, ?>, Foo<u64, bool, ?>]
+    //     [Foo<u8, ?, ?>, Foo<u64, bool, ?>, Bar]
+    //     [Foo<u8, ?, ?>, Foo<u64, bool, Bar>]
+    //     [Foo<u8, Foo<u64, bool, Bar>, ?>]
+    //     [Foo<u8, Foo<u64, bool, Bar>, ?>, address]
+    //     [Foo<u8, Foo<u64, bool, Bar>, address>]        (done)
+
+    use SerializedType as S;
+
+    enum TypeBuilder {
+        Saturated(SignatureToken),
+        Vector,
+        Reference,
+        MutableReference,
+        StructInst {
+            sh_idx: StructHandleIndex,
+            arity: usize,
+            ty_args: Vec<SignatureToken>,
+        },
+    }
+
+    impl TypeBuilder {
+        fn apply(self, tok: SignatureToken) -> Self {
+            match self {
+                T::Vector => T::Saturated(SignatureToken::Vector(Box::new(tok))),
+                T::Reference => T::Saturated(SignatureToken::Reference(Box::new(tok))),
+                T::MutableReference => {
+                    T::Saturated(SignatureToken::MutableReference(Box::new(tok)))
+                }
+                T::StructInst {
+                    sh_idx,
+                    arity,
+                    mut ty_args,
+                } => {
+                    ty_args.push(tok);
+                    if ty_args.len() >= arity {
+                        T::Saturated(SignatureToken::StructInstantiation(sh_idx, ty_args))
+                    } else {
+                        T::StructInst {
+                            sh_idx,
+                            arity,
+                            ty_args,
+                        }
+                    }
+                }
+                _ => unreachable!("invalid type constructor application"),
             }
-            SerializedType::STRUCT_INST => {
-                let sh_idx = read_uleb_u16_internal(cursor)?;
-                let type_params = load_signature_tokens(cursor)?;
-                Ok(SignatureToken::StructInstantiation(
-                    StructHandleIndex(sh_idx),
-                    type_params,
-                ))
+        }
+
+        fn is_saturated(&self) -> bool {
+            matches!(self, T::Saturated(_))
+        }
+
+        fn unwrap_saturated(self) -> SignatureToken {
+            match self {
+                T::Saturated(tok) => tok,
+                _ => unreachable!("cannot unwrap unsaturated type constructor"),
```

Comment:
this is always a bit scary to me :p since this function is only used in one place, can we just inline this?

(sorry for the super late review haha)